src/C-interface/ellpack/bml_multiply_ellpack_typed.c

```diff
@@ -1272,4 +1288,185 @@ void TYPED_FUNC(
+
+    REAL_T threshold = (REAL_T) threshold1;
+
+    // convert ellpack to cucsr
+    TYPED_FUNC(bml_ellpack2cucsr_ellpack) (A);
+    TYPED_FUNC(bml_ellpack2cucsr_ellpack) (B);
+    TYPED_FUNC(bml_ellpack2cucsr_ellpack) (C);
+
+    // Create hypre csr matrices A and B
+    // Note: The following update is not necessary since the ellpack2cucsr
+    // routine updates the csr rowpointers on host and device
+//#pragma omp target update from(csrRowPtrA[:A_N+1])
+//#pragma omp target update from(csrRowPtrB[:B_N+1])
+//#pragma omp target update from(csrRowPtrC[:C_N+1])
+    int nnzA = csrRowPtrA[A_N];
+    int nnzB = csrRowPtrB[B_N];
+    int nnzC_in = csrRowPtrC[C_N];
+
+//     HYPRE_Init();    
+//     HYPRE_SetExecutionPolicy(HYPRE_EXEC_DEVICE);
+     int use_vendor = 0;
+     int spgemm_alg = 1;
+     int spgemm_binned = 0;
+     HYPRE_SetSpGemmUseVendor(use_vendor);
+     hypre_SetSpGemmAlgorithm(spgemm_alg);
+     hypre_SetSpGemmBinned(spgemm_binned);
+    /* create hypre csr matrix */
+    matA = hypre_CSRMatrixCreate( A_N,A_N,nnzA );
+    matB = hypre_CSRMatrixCreate( B_N,B_N,nnzB );
+    matC = hypre_CSRMatrixCreate( C_N,C_N,nnzC_in );
+
+#pragma omp target data use_device_ptr(csrRowPtrA,csrColIndA,csrValA, \
+		csrRowPtrB,csrColIndB,csrValB, \
+		csrRowPtrC,csrColIndC,csrValC)
+    {
+       hypre_CSRMatrixI(matA) = csrRowPtrA;
+       hypre_CSRMatrixJ(matA) = csrColIndA;
+       hypre_CSRMatrixData(matA) = csrValA;
+
+       hypre_CSRMatrixI(matB) = csrRowPtrB;
+       hypre_CSRMatrixJ(matB) = csrColIndB;
+       hypre_CSRMatrixData(matB) = csrValB;
+
+       hypre_CSRMatrixI(matC) = csrRowPtrC;
+       hypre_CSRMatrixJ(matC) = csrColIndC;
+       hypre_CSRMatrixData(matC) = csrValC;
+    }
+ 
+    hypre_CSRMatrix *matD  = hypre_CSRMatrixMultiplyDevice(matA, matB);
+
+    /* add matrices */
+    int spadd_use_vendor=0;
+    HYPRE_SetSpAddUseVendor(spadd_use_vendor);
+    hypre_SetSpAddAlgorithm(1);
+    hypre_CSRMatrix *matE = hypre_CSRMatrixAddDevice(alpha, matD, beta, matC);
+
+        // Place the resulting matrix in C
+    if (is_above_threshold(threshold, BML_REAL_MIN))
+    {
+       int nnzE = hypre_CSRMatrixNumNonzeros(matE);
+       REAL_T *elmt_tol =
+               (REAL_T *) malloc(sizeof(REAL_T) * nnzE);
+        // Allocate the working arrays on the device
+#pragma omp target enter data map(alloc:elmt_tol[:nnzE])
+
+#pragma omp target teams distribute parallel for
+        for(int i = 0; i<nnzE; i++) {
+          elmt_tol[i] = threshold;
+        }
+#pragma omp target data use_device_ptr(elmt_tol)
+        {
+        hypre_CSRMatrixDropSmallEntriesDevice( matE, threshold, elmt_tol);
+        }
+
+#pragma omp target exit data map(delete:elmt_tol[:hypre_CSRMatrixNumNonzeros(matE)])
+free(elmt_tol);
+
+    }    
+
+    // Done with matrix multiplication.
+    // Update ellpack C matrix (on device): copy from csr to ellpack format
+/*
+#pragma omp target data use_device_ptr(csrRowPtrC,csrColIndC,csrValC)
```

Comment:
clean up comments

Reply:
This commented section is actually what is needed for AMD. I left that in there as a place holder for now. I have updated the comments to reflect this.